File: src/solutions/2023/02/lua/main.lua
local run = require('src.langs.lua.run')
local str = require('src.langs.lua.utils.str')

local options = {
   has_alternate = false,
   has_io = false,
}

local function parse_input(input)
   local results = {}
   for line in input:gmatch('[^\n]+') do
      local cubes = {}
      for _, c in ipairs(str.split(str.split(line, ':')[2]:gsub(',', ';'), ';')) do
         local lmao = str.split(c:gsub('^%s*(.-)%s*$', '%1'), ' ')
         local val = tonumber(lmao[1])
         local col = (lmao[2]:sub(1, 1):byte() % 3) + 1
         cubes[#cubes + 1] = { val, col }
      end

      results[#results + 1] = cubes
   end

   return results
end

local rgb_p1 = { 12, 13, 14 }

---@param input string
---@param is_test boolean
---@return string
local function part_1(input, is_test)
   local parsed = parse_input(input)

   local res = 0
   for i = 1, #parsed do
      local cubes = parsed[i]
      local found = false
      for j = 1, #cubes do
         local cube = cubes[j]
         if cube[1] > rgb_p1[cube[2]] then
            found = true
         end
      end
      if not found then
         res = res + i
      end
   end
   return tostring(res)
end

---@param input string
---@param is_test boolean
---@return string
local function part_2(input, is_test)
   local parsed = parse_input(input)

   local res = 0
   for i = 1, #parsed do
      local cubes = parsed[i]
      local rgb = { 0, 0, 0 }
      for j = 1, #cubes do
         local cube = cubes[j]
         rgb[cube[2]] = math.max(rgb[cube[2]], cube[1])
      end

      res = res + rgb[1] * rgb[2] * rgb[3]
   end
   return tostring(res)
end

local function main()
   run(arg, part_1, part_2, options)
end

main()

File: src/solutions/2023/02/zig/main.zig
const std = @import("std");
const aoc = @import("aoc");

const options = aoc.SolutionOptions.init();

const pogcube = struct { value: u64, color: u8 };

fn parseInput(input: []const u8) [][]pogcube {
    var lines = std.mem.splitSequence(u8, input, "\n");
    var results = std.ArrayList([]pogcube).init(std.heap.page_allocator);
    while (lines.next()) |line| {
        var x = std.mem.splitBackwardsSequence(u8, line, ":");
        const l = x.next().?;
        const m = std.mem.replaceOwned(u8, std.heap.page_allocator, l, ",", ";") catch "";
        var poggy = std.mem.splitSequence(u8, m, ";");
        var cubes = std.ArrayList(pogcube).init(std.heap.page_allocator);
        while (poggy.next()) |poggo| {
            const why = std.mem.trim(u8, poggo, " ");
            var pog = std.mem.splitSequence(u8, why, " ");
            const val = std.fmt.parseInt(u64, pog.next().?, 10) catch unreachable;
            const col = @as(u8, @mod(pog.next().?[0], 3));
            cubes.append(.{ .value = val, .color = col }) catch unreachable;
        }
        results.append(cubes.toOwnedSlice() catch unreachable) catch unreachable;
    }

    return results.toOwnedSlice() catch unreachable;
}

const rgb_p1: [3]u64 = .{ 12, 13, 14 };
fn part1(input: []const u8, is_test: bool) []const u8 {
    _ = is_test;
    const parsed = parseInput(input);
    defer std.heap.page_allocator.free(parsed);
    var res: u64 = 0;

    for (parsed, 0..) |cubes, i| {
        defer std.heap.page_allocator.free(cubes);
        var found = false;
        for (cubes) |cube| {
            if (cube.value > rgb_p1[cube.color]) {
                found = true;
                break;
            }
        }
        if (!found) {
            res = res + i + 1;
        }
    }

    const buf = std.heap.page_allocator.alloc(u8, 42) catch unreachable;
    const str = std.fmt.bufPrint(buf, "{}", .{res}) catch "";
    return str;
}

fn part2(input: []const u8, is_test: bool) []const u8 {
    _ = is_test;
    const parsed = parseInput(input);
    defer std.heap.page_allocator.free(parsed);
    var res: u64 = 0;

    for (parsed) |cubes| {
        defer std.heap.page_allocator.free(cubes);
        var rgb: [3]u64 = .{ 0, 0, 0 };
        for (cubes) |cube| {
            if (cube.value > rgb[cube.color]) {
                rgb[cube.color] = cube.value;
            }
        }
        res = res + rgb[0] * rgb[1] * rgb[2];
    }

    const buf = std.heap.page_allocator.alloc(u8, 42) catch unreachable;
    const str = std.fmt.bufPrint(buf, "{}", .{res}) catch "";
    return str;
}

pub fn main() !void {
    try aoc.run(part1, part2, options);
}
